docs/Developers/Design/GD2_node_state_machine.md

```diff
@@ -0,0 +1,131 @@
+it will not be used for new allocations, nor does it hold any user data. Once
+the operator sees the node has reached this state, the operator knows it is
+safe to delete the node. The operator removes the node from the TSP, followed
+by deleting the Deployment/pod for the node and freeing the storage that backed
+the node.
+
+# Node maintenance
+
+It is necessary to notify the operator and GD2 when troubleshooting or
+performing maintenance either on the Gluster cluster, or on infrastructure,
+such that it may affect the availability of a Gluster node. The reason for
+notification is to prevent the system (operator + GD2) from taking corrective
+action if they detect the node is down and to prevent them from making other
+changes to the node that may interfere with the ongoing maintenance.
+
+Maintenance is signaled by moving the node into the **quiesced** state. A node
+may enter this state from either **online** or **offline** (depending on
+whether the node is currently up or down). The transition would be requested
+via the operator. Once a node has entered the **quiesced** state, GD2 would not
+trigger any migrations nor allocations on this node. It also would not mark a
+node as **failed** if it is down for an extended period while in this state.
+
+Once maintenance is completed, the operator can take the node out of
+"maintenance" by triggering the transition to either the **online** or
+**offline** state, depending on whether GD2 considers the node up or down.
+(again, this is slightly simplified for explanation)
+
+# State machine implementation
+
+In the above description, some simplifications were made. For example, the
+quiesced state has inconsistencies: When the operator triggers the exit from
+the quiesced state, it cannot directly set the new state because online/offline
+is a distiction that can only be made by GD2.
+
+While it is possible to resolve these issues by adding additional states and
+transitions (e.g., quiescedDown and quiescedUp), another approach is to
+represent the state machine as a set of boolean conditions that can be changed
+by either the operator or GD2:
+
+- **isUp** - (GD2) - True if the node is up according to GD2
+- **hasData** - (GD2) - True if the node contains 1 or more bricks
+- **shouldQuiesce** - (operator) - True if the node should be quiesced
+- **shouldDestroy** - (GD2/operator) - True if the node should be destroyed
+
+By toggling the appropriate conditions, above, it is possible to implement the
+state machine:
+
+Condition     | online | offline | quiesced | failed | abandoned
+--------------|:------:|:-------:|:--------:|:------:|:--------:
+isUp          | 1      | 0       | x        | x      | x
+hasData       | x      | x       | x        | 1      | 0
+shouldQuiesce | 0      | 0       | 1        | x      | x
+shouldDestroy | 0      | 0       | 0        | 1      | 1
+
+Note:
+
+- All 16 combinations are represented in the table, above
```

Comment:
Its 16 or 20 combinations?

Reply:
There are 4 binary flags (condition column) which leads to 2^4 = 16 combinations. These 16 combinations are then mapped into the 5 "states" that the document discusses.